botocore/client.py

```diff
@@ -175,14 +180,38 @@ def _get_s3_addressing_handler(self, endpoint_url, s3_config,
         return fix_s3_host
 
     def _is_s3_accelerate(self, endpoint_url, s3_config):
+        # Accelerate has been explicitly configured
         if s3_config is not None and s3_config.get('use_accelerate_endpoint'):
             return True
 
+        # Accelerate mode is turned on automatically if an endpoint url is
+        # provided that matches the accelerate scheme.
         if endpoint_url is None:
             return False
 
-        parts = urlsplit(endpoint_url).netloc.split('.')
-        return any(part == 's3-accelerate' for part in parts)
+        # Accelerate is only valid for Amazon endpoints.
+        netloc = urlsplit(endpoint_url).netloc
+        if not netloc.endswith('amazonaws.com'):
+            return False
+
+        # The first part of the url should always be s3-accelerate
```

Comment:
That wouldn't weed out duplicates, and it's not very clear why you're testing for 3 and 4. Actually, is there any reason to constrain based on length? So long as we make sure that each part is in the whitelist and there are no duplicates we'll catch everything anyway and that will remove the need to update this if the url feature matrix gets bigger.